docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,127 @@
+* To assign a VIP, we build an IPAM state reserving every address we already assigned and allocate the next address.
+  To solve the problem of “assigning immediately” a VIP of an old service to a new Service we can also store in memory recently removed VIPs for X seconds and set TTL of DNS responses for X seconds.
+  The current value of 30s is too long, especially that the DNS server is run locally in kuma-dp. Leader election should take more than X seconds, so if we were to lose this list, it’s not a problem.
+
+Following this process, we don’t have to store VIPs in any central place like we do currently (VIP ConfigMap for each Mesh).
+We also assign VIPs for ExternalServices, but this can work as long as we consider both ExternalServices and MeshServices for this process.
+
+To avoid collision with an existing VIPs mechanism, we can take a completely new CIDR by default (241.0.0.0/4).
+This CIDR can be configured by the user.
+
+### Hostname Generator
+
+MeshService has to be addressable with a hostname.
+In case of services in a Kubernetes cluster, we don't have to do anything because we can just rely on Kube DNS that resolves to a Kube VIP we placed for MeshService.
+
+In case of services in a Universal cluster, currently we provide two ways:
+* `<kuma.io/service>.mesh` address that resolves to Kuma VIP
+* VirtualOutbound resource to provide custom addresses
+
+Hostnames should be managed only by MeshOperator. Otherwise, we get into security issues such as one service owner can override a hostname of other service.
+Hostnames are created by convention in the majority of cases - Kubernetes hostnames are also such example.
+It is still possible for service owner to create unique hostname for their app by reaching out to mesh operator.
+
+To manage hostname, we introduce new object called `HostnameGenerator`
+
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
+spec:
+  meshServiceSelector:
+    matchLabels:
+      kuma.io/zone: east 
+  template: {{ label "k8s.kuma.io/service-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.cluster.{{ label "kuma.io/zone" }}
+```
+
+`HostnameGenerator` is a namespaced-scoped object on Kubernetes, but can only be applied in system namespace.
+It can be applied on global CP, so it's synced to all zones, or it can be applied on specific zones.
+
+For example, given this mesh service
+
+```yaml
+type: MeshService
+name: redis.demo-app
+labels:
+  k8s.kuma.io/service-name: redis
+  k8s.kuma.io/namespace: demo-app
+  kuma.io/zone: east
+spec: ...
+```
+We would generate such hostname `redis.demo-app.svc.cluster.east`
+
+Possible template functions/keys are:
+* `{{ name }}` - name of the MeshService
+* `{{ label "x" }}` - value of label `x`.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a `NotAvailable` status and an appropriate `reason`.
```

Comment:
I guess that's on the MeshService? Shouldn't the status be more a `HostnameNotAvailable` or something like this? Feels like NotAvailable is a bit too generic.
I guess we also send an event right?

Reply:
Ok got confused because status is explained after.

Reply:
Should there be a Condition on MeshService?

Reply:
Hard to say. I did some reading on conditions https://maelvls.dev/kubernetes-conditions/
To try to convert this to conditions it would be something like
```
status:
  conditions:
  - type: HostnameAssigned
    status: "False"
    message: "There is a conflict..."
```
this `HostnameAssigned` would be for all hostnames. Now the questions are
1) What to do when there is one hostname and second cannot be assigned because of conflict. Is it True or False?
2) What to do when we have a Kube service with kube hostname. We don't put Kube DNS hostname in list of hostnames, so is it True or False?

My feeling is that the status of each hostname is more useful. Unless we push conditions into each item of `addresses`, but sounds like a simple status & reason would be simpler API?

Reply:
It could be `AllHostnameAssigned` no?

Reply:
and if no hostnames then it's `True`?

Reply:
talked offline: let's start with status and then we can do conditions if needed